ray-operator/controllers/ray/expectations/active_expectation.go

```diff
@@ -0,0 +1,382 @@
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get active expectation item for %s when expecting: %s", key, err)
+	}
+
+	ae.recordTimestamp = time.Now()
+	if !exist {
+		item = &ActiveExpectationItem{Client: ae.Client, Name: name, Kind: kind, Key: key, Action: action, RecordTimestamp: time.Now()}
+		if err := ae.items.Add(item); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	item.(*ActiveExpectationItem).Action = action
+	item.(*ActiveExpectationItem).RecordTimestamp = time.Now()
+	return nil
+}
+
+func (ae *ActiveExpectation) expectUpdate(kind ExpectedResourceType, name string, resourceVersion int64) error {
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get active expectation item for %s when expecting: %s", key, err)
+	}
+
+	ae.recordTimestamp = time.Now()
+	if !exist {
+		item = &ActiveExpectationItem{Client: ae.Client, Name: name, Kind: kind, Key: key, Action: Update, ResourceVersion: resourceVersion, RecordTimestamp: time.Now()}
+		if err := ae.items.Add(item); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	ea := item.(*ActiveExpectationItem)
+	ea.Action = Update
+	ea.ResourceVersion = resourceVersion
+	ea.RecordTimestamp = time.Now()
+
+	return nil
+}
+
+func (ae *ActiveExpectation) isSatisfied() (satisfied bool, err error) {
+	items := ae.items.List()
+
+	satisfied = true
+	for _, item := range items {
+		itemSatisfied, itemErr := func() (satisfied bool, err error) {
+			defer func() {
+				if satisfied {
+					ae.items.Delete(item)
+				} else if ae.recordTimestamp.Add(ExpectationsTimeout).Before(time.Now()) {
+					panic("expected panic for active expectation")
+				}
+			}()
+
+			satisfied, err = item.(*ActiveExpectationItem).isSatisfied(ae.namespace)
+			if err != nil {
+				return false, err
+			}
+
+			return satisfied, nil
+		}()
+
+		if itemErr != nil && err == nil {
+			err = fmt.Errorf("fail to check satisfication for subject %s, item %s: %s", ae.key, item.(*ActiveExpectationItem).Key, itemErr)
+		}
+
+		satisfied = satisfied && itemSatisfied
+	}
+
+	return satisfied, err
+}
+
+func (ae *ActiveExpectation) delete(kind, name string) error {
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to delete active expectation item for %s: %s", key, err)
+	}
+
+	if !exist {
+		return nil
+	}
+
+	if err := ae.items.Delete(item); err != nil {
+		return fmt.Errorf("fail to do delete active expectation item for %s: %s", key, err)
+	}
+
+	return nil
+}
+
+type ActiveExpectationItem struct {
+	client.Client
+
+	Key             string
+	Name            string
+	Kind            ExpectedResourceType
+	Action          ActiveExpectationAction
+	ResourceVersion int64
+	RecordTimestamp time.Time
+}
+
+func (i *ActiveExpectationItem) isSatisfied(namespace string) (bool, error) {
```

Comment:
Do we return any non-nil `error` in the function?

Reply:
fixed. Deleted error return.